### PR TITLE
Farm Staking SC

### PR DESCRIPTION
**Pull request by dorin-iancu**

### CODE REVIEW

File: dex/farm/src/lib.rs [hunk 1]
```diff
@@ -229,11 +248,63 @@ pub trait Farm:
...
+        #[var_args] opt_accept_funds_func: OptionalArg<ManagedBuffer>,
+    ) -> SCResult<()> {
+        require!(self.is_active(), "Not active");
+        require!(!self.farm_token_id().is_empty(), "No farm token");
+
+        let farm_token_id = self.farm_token_id().get();
+        require!(payment_token_id == farm_token_id, "Bad input token");
+        require!(amount > 0, "Payment amount cannot be zero");
+
+        self.farming_token_total_liquidity()
```
Comment (sasurobert) [inline dex/farm/src/lib.rs:272]: this has to be moved to exitFarm -> unStakeFarm method. there are no rewards between unStake and unBond.

File: dex/farm/src/lib.rs [hunk 2]
```diff
@@ -229,11 +248,63 @@ pub trait Farm:
...
+            &farm_token_id,
+            token_nonce,
+        );
+        let full_unlock_epoch = token_info
+            .decode_attributes::<UnbondSftAttributes>()?
+            .full_unlock_epoch;
+        let current_epoch = self.blockchain().get_block_epoch();
+
+        let farming_token_id = self.farming_token_id().get();
+        let farming_token_amount = if current_epoch < full_unlock_epoch {
```
Comment (sasurobert) [inline dex/farm/src/lib.rs:286]: let's remove penalty completely. this is not needed in staking contract

File: dex/farm/src/lib.rs [hunk 3]
```diff
@@ -107,9 +112,12 @@ pub trait Farm:
         require!(token_in == farming_token_id, "Bad input token");
         require!(enter_amount > 0, "Cannot farm with amount of 0");
 
+        self.farming_token_total_liquidity()
```
Comment (sasurobert) [inline dex/farm/src/lib.rs:115]: why is this needed ? Where is this inside the farm contracts ?

Comment (sasurobert) [reply dex/farm/src/lib.rs:115]: oh, I see. For the bounded APR computation

File: dex/farm/src/lib.rs (hunk 2 above)
Comment (dorin-iancu) [reply dex/farm/src/lib.rs]: Done. Now the unbond period is forced, with no early unbond allowed.

File: dex/farm/src/lib.rs (hunk 1 above)
Comment (dorin-iancu) [reply dex/farm/src/lib.rs]: Done.

File: dex/farm/src/lib.rs (hunk 3 above)
Comment (dorin-iancu) [reply dex/farm/src/lib.rs:115]: Yep. It's the only reason it's needed. Couldn't use SC balance directly in case of farming token being the same as the reward token.

File: dex/farm/src/custom_rewards.rs [hunk 4]
```diff
@@ -1,54 +1,121 @@
...
+    ) -> BigUint {
+        let unbounded_rewards =
+            self.calculate_reward(amount, current_reward_per_share, initial_reward_per_share);
+        if unbounded_rewards == 0u32 {
+            return unbounded_rewards;
+        }
+
+        let farming_token_total_liquidity = self.farming_token_total_liquidity().get();
+        let max_apr = self.max_annual_percentage_rewards().get();
+        let max_rewards = farming_token_total_liquidity * max_apr
```
Comment (catalinnnn) [inline dex/farm/src/custom_rewards.rs:50]: not sure about this calculus :thinking: maybe we'll reconsider this together

Comment (dorin-iancu) [reply dex/farm/src/custom_rewards.rs]: Fixed in latest commit.
